src/spatial_data_format.rs

```diff
@@ -0,0 +1,558 @@
+    fn streams_quantized(&self, sample_rate: u64) -> Vec<(GrapeTag, Vec<f32>)> {
+        debug_assert!(sample_rate < self.header.sample_rate);
+        let pts_per_sample = self.header.sample_rate as usize / sample_rate as usize;
+        let raw_streams = self.get_raw_streams();
+        let quantized_streams = raw_streams
+            .into_iter()
+            .map(|v| {
+                v.chunks(pts_per_sample)
+                    .map(|c| c.iter().sum::<f32>() / c.len() as f32)
+                    .collect()
+            })
+            .collect();
+        Self::attach_tags(&self.header.tags, quantized_streams)
+    }
+}
+
+/// This builder contains the data required
+#[derive(Debug, Clone)]
+pub struct GrapeFileBuilder {
+    sample_rate: u64,
+    streams: Vec<(GrapeTag, Vec<f32>)>,
+}
+
+impl Default for GrapeFileBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GrapeFileBuilder {
+    /// Instaintiate a builder with no streams and a default sample rate of
+    /// 1000 samples per second.
+    fn new() -> Self {
+        GrapeFileBuilder {
+            sample_rate: 1000,
+            streams: Vec::new(),
+        }
+    }
+
+    /// Sets the sample rate of the builder to the argument.
+    pub fn set_samplerate(self, sample_rate: u64) -> Self {
+        GrapeFileBuilder {
+            sample_rate,
+            ..self
+        }
+    }
+
+    /// Adds a tagged stream to the builder.
+    pub fn add_stream(mut self, stream: &[f32], tag: GrapeTag) -> Self {
+        let stream: Vec<f32> = stream.to_vec();
+        self.streams.push((tag, stream));
+        self
+    }
+
+    /// Removes all streams from the builder
+    pub fn clear_streams(mut self) -> Self {
+        self.streams.clear();
+        self
+    }
+
+    /// Builds a [GrapeFile] from the builder, truncating all streams to the
+    /// length of the shortest stream.
+    pub fn build_truncate(self) -> GrapeFile {
+        let tags: Vec<GrapeTag> = self
+            .streams
+            .iter()
+            .map(|(tag, _vec)| tag)
+            .cloned()
+            .collect();
+        let sample_vecs: Vec<Vec<f32>> = self.streams.into_iter().map(|(_tag, vec)| vec).collect();
+
+        let mut samples = Vec::new();
+
+        let shortest = sample_vecs.iter().map(|v| v.len()).min();
+        if let Some(shortest) = shortest {
+            samples.reserve_exact(shortest * sample_vecs.len() as usize);
+            for sample_idx in 0..shortest {
+                for stream_idx in 0..sample_vecs.len() {
+                    samples.push(sample_vecs[stream_idx as usize][sample_idx]);
+                }
+            }
+        };
+
+        GrapeFile {
+            header: GrapeFileHeader {
+                n_streams: sample_vecs.len() as u64,
+                sample_rate: self.sample_rate,
+                tags,
+            },
+            samples,
+        }
+    }
+
+    /// Builds a [GrapeFile] from the builder, extending all streams with the
```

Comment:
This implementation of `build_extend` extends the last value of short streams.

Reply:
Yes, I know! I personally think it may be better to fill it with silence rather than extending the last value, so a mixed audio file would fade out in the case of unequal stream length rather than sounding full of junk. This implementation is also fine, I only wanted to bring it up as a possible discussion point :)

Reply:
Oh, sorry, I misunderstood! I lean towards extending the last value because these aren't really audio channels, they represent positional movement. I think it's okay to say "if you stop sending positional data, I'll assume you stopped moving". What do you think?

Reply:
Ohhhh doh. You are so right. Better to keep last position than rubberband back to center. Sounds good :)